app/src/org/commcare/android/resource/AndroidResourceManager.java

```diff
@@ -0,0 +1,258 @@
+        tempUpgradeTable.setInstallCancellationChecker(cancelCheckker);
+    }
+
+    /**
+     * Save upgrade stats if the upgrade was cancelled and wasn't complete at
+     * that time.
+     */
+    public void upgradeCancelled() {
+        if (!isUpgradeTableStaged()) {
+            UpdateStats.saveStatsPersistently(app, updateStats);
+        } else {
+            Log.i(TAG, "Upgrade cancelled, but already finished with these stats");
+            Log.i(TAG, updateStats.toString());
+        }
+    }
+
+    public void incrementUpdateAttempts() {
+        updateStats.registerStagingAttempt();
+    }
+
+    /**
+     * Log update failure that occurs while trying to install the staged update table
+     */
+    public void recordUpdateInstallFailure(Exception exception) {
+        updateStats.registerUpdateException(exception);
+    }
+
+    public void recordUpdateInstallFailure(AppInstallStatus result) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+    }
+
+    /**
+     * Clear update table, log failure with update stats,
+     * and, if appropriate, schedule a update retry
+     *
+     * @param result update attempt result
+     * @param ctx    Used for showing pinned notification of update task retry
+     */
+    public void processUpdateFailure(AppInstallStatus result, Context ctx) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+
+        boolean reusePartialTable =
+                (result == AppInstallStatus.UnknownFailure ||
+                        result == AppInstallStatus.NoLocalStorage);
+
+        if (!reusePartialTable) {
+            upgradeTable.clear();
+        }
+
+        retryUpdateOrGiveUp(ctx);
+    }
+
+    private void retryUpdateOrGiveUp(Context ctx) {
+        if (updateStats.isUpgradeStale()) {
+            Log.i(TAG, "Stop trying to download update. Here are the update stats:");
+            Log.i(TAG, updateStats.toString());
+
+            UpdateStats.clearPersistedStats(app);
+
+            upgradeTable.clear();
+        } else {
+            Log.w(TAG, "Retrying auto-update");
+            UpdateStats.saveStatsPersistently(app, updateStats);
+            scheduleUpdateTaskRetry(ctx);
+        }
+    }
+
+    private void scheduleUpdateTaskRetry(final Context ctx) {
+        final Handler handler = new Handler();
+        handler.postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                String ref = ResourceInstallUtils.getDefaultProfileRef();
+                try {
+                    UpdateTask updateTask = UpdateTask.getNewInstance();
+                    updateTask.startPinnedNotification(ctx);
+                    updateTask.execute(ref);
+                } catch (IllegalStateException e) {
+                    // The user may have started the update process in the meantime
+                    Log.w(TAG, "Trying trigger an auto-update retry when it is already running");
```

Comment:
hm, wondering whether if there is some stateful stuff here that something should be locking, or will that always come through on the execute?